### PR TITLE
Fix subnet overlapping

### LINKED ISSUE #982: VPC subnet calculation for small CIDR ranges fails
### What happened?

I've defined a VPC with a `/22` cidr range

```
vpc = awsx.ec2.Vpc(
    f"vpc-{stack_name}",
    cidr_block="172.16.0.0/22",
    subnet_specs=[
        awsx.ec2.SubnetSpecArgs(
            type="public",
        ),
        awsx.ec2.SubnetSpecArgs(
            type="private",
        ),
    ],
    number_of_availability_zones=2,
    nat_gateways=awsx.ec2.NatGatewayConfigurationArgs(strategy=awsx.ec2.NatGatewayStrategy.ONE_PER_AZ),
)
```

No matter what I do to manipulate the cidr masks, I still can't get a valid VPC 

### Expected Behavior

In awsx classic, it was possible to have many different VPC configurations. I expect awsx 1.0.0 to work that way

### Steps to reproduce

Define a VPC with a cidr smaller than a `/20

### Output of `pulumi about`

Not necessary

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by danielrbradley**
Fix a bug in the subnet distributor which causes CIDR blocks to be generated where the blocks overlap with one another - causing an undeployable configuration.

Generate 100 test cases of known working subnet configurations to ensure we don't change the ouput for existing deployments. We **only** want to change the behaviour for cases which current cause overlaps.

Combine the old broken allocation method with an additional check for overlapping the previous block. If there's an overlap, then use the new strategy. This avoids affecting anyone who's already got a working deployment. Using only the new method sometimes generates more compact arrangements hence causing breaking changes.

- This is stacked on top of https://github.com/pulumi/pulumi-awsx/pull/1130
- Fixes #982

### CODE REVIEW

File: awsx/ec2/subnetDistributor.ts [hunk 1]
```diff
@@ -126,17 +156,22 @@ export function getSubnetSpecs(
         splitBase = azBases[i];
       }
 
-      const isolatedCidrMask: number = isolatedSubnetsIn[j].cidrMask ?? 24;
+      const isolatedCidrMask: number =
+        isolatedSubnetsIn[j].cidrMask ?? Math.max(baseSubnetMask + newBitsPerSubnet, 24);
       const newBits = isolatedCidrMask - baseIp.subnetMask;
+      let nextAddress = new ipAddress.Address4(cidrSubnetV4(splitBase, newBits, j));
+      if (isOverlapping(currentAddress, nextAddress)) {
+        nextAddress = nextBlock(currentAddress, isolatedCidrMask);
```
Comment (VenelinMartinov) [inline awsx/ec2/subnetDistributor.ts:164]: Shouldn't this increment nextAddress, not currentAddress? Also is it guaranteed that the new nextAddress does not overlap with currentAddress?

File: awsx/ec2/subnetDistributor.ts [hunk 2]
```diff
@@ -201,6 +235,55 @@ function cidrSubnetV4(ipRange: string, newBits: number, netNum: number): string
...
+export function validateRanges(specs: SubnetSpec[]) {
+  const ranges = specs
+    .map((spec) => new ipAddress.Address4(spec.cidrBlock))
+    .map((addr) => {
+      const start = addr.startAddress().bigInteger();
+      const end = addr.endAddress().bigInteger();
+      return { addr, start, end };
+    });
+  let previous = ranges[0];
+  for (const current of ranges.slice(1)) {
```
Comment (VenelinMartinov) [inline awsx/ec2/subnetDistributor.ts:247]: Is it possible for addresses to overlap one apart? ex. index 1 with index 3

File: awsx/ec2/subnetDistributor.test.ts [hunk 3]
```diff
@@ -12,8 +12,168 @@
...
+          .record({
+            vpcCidrMask: cidrMask(),
+            // We're just focusing on the logic of choosing the next range, so we don't need large numbers of specs.
+            // Similarly, we only need a single AZ because AZs are always split evenly.
+            subnetSpecs: fc.array(subnetSpec(), { minLength: 1, maxLength: 2 }),
+          })
+          .filter(({ vpcCidrMask, subnetSpecs }) => {
+            for (const subnetSpec of subnetSpecs) {
+              // Requite subnet to be smaller than VPC
+              if (subnetSpec.cidrMask! <= vpcCidrMask) {
```
Comment (VenelinMartinov) [inline awsx/ec2/subnetDistributor.test.ts:138]: What does ! do here?

Comment (danielrbradley) [reply awsx/ec2/subnetDistributor.test.ts:138]: It asserts that cidrMask is not undefined. The type of cidrMask is `number | undefined`, but we're always specifying a value for it in the setup, so using `!` is fine in a test ... which makes the numeric comparison easier to write rather than having to have a separate expression to prove to the compiller that it's not undefined.

File: awsx/ec2/subnetDistributor.ts (hunk 1 above)
Comment (danielrbradley) [reply awsx/ec2/subnetDistributor.ts:164]: Yup, it does - it firstly overwrites the nextAddress because the old nextAddress was overlapping, then on line 174 we allocate the nextAddress back to the currentAddress. I've used this layout in each looop.

Yes, the method used in nextBlock is guarenteed to not overlap.

File: awsx/ec2/subnetDistributor.ts (hunk 2 above)
Comment (danielrbradley) [reply awsx/ec2/subnetDistributor.ts:247]: All the allocations are done incrementally so we're just finding the next based on the previous so I think we're okay. Though one edge case is if the user specified masks significantly exceed the space available within the first AZ, causing it to massively overlap into the next AZ. I don't think we need to worry about that for our tests yet though.

**Review by VenelinMartinov — APPROVED** [review]
